pushflatpakscript/src/pushflatpakscript/flathub.py

```diff
@@ -104,20 +104,29 @@ def check_and_extract_tar_archive(context, tar_file_path):
             log.warning(f"{tar_file_path} does not have `repo` as topdir")
             flatpak_deflated_dir = topdir
 
-        def is_within_directory(directory, target):
-            abs_directory = os.path.abspath(directory)
-            abs_target = os.path.abspath(target)
-            prefix = os.path.commonprefix([abs_directory, abs_target])
-            return prefix == abs_directory
-
         def safe_extract(tar, path=".", members=None, *, numeric_owner=False):
-            for member in tar.getmembers():
+            def is_within_directory(directory, target):
+                real_directory = os.path.realpath(directory)
+                real_target = os.path.realpath(target)
+                prefix = os.path.commonprefix([real_directory, real_target])
+                return prefix == real_directory
+
+            def validate_tar_member(member, path):
                 member_path = os.path.join(path, member.name)
                 if not is_within_directory(path, member_path):
                     raise Exception("Attempted path traversal in tar file: " + member.name)
+                if member.issym():
+                    link_path = os.path.join(os.path.dirname(member_path), member.linkname)
+                    if not is_within_directory(path, link_path):
+                        raise Exception("Attempted link path traversal in tar file: " + member.name)
                 if member.mode & (stat.S_ISUID | stat.S_ISGID):
                     raise Exception("Attempted setuid or setgid in tar file: " + member.name)
-            tar.extractall(path, members, numeric_owner=numeric_owner)
+            def files(tar, path):
+                for member in tar:
+                    validate_tar_member(member, path)
+                    yield member
+
+            tar.extractall(path, members=files(tar, path), numeric_owner=numeric_owner)
```

Comment:
We should either remove the `members` param from safe_extract, or use it.  (AFAICT the single caller doesn't pass it, so removing is maybe the easier/saner option.)

Reply:
Thanks.